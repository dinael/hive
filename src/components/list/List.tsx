import { FC } from 'react'
import { v4 as uuid } from 'uuid'

import styles from './List.module.scss'

export type ListItem = {
  id?: string | number,
  text: string
}

export type ListProps = {
  items: ListItem[]
  ordered?: boolean
  nameList?: string
  className?: string
  disabled?: boolean
  unstyled?: boolean
  bullet?: 'none' | 'auto' | 'disc' | 'circle' | 'square' | 'decimal' | 'decimal-leading-zero' | 'lower-roman' | 'upper-roman' | 'lower-greek' | 'lower-latin' | 'upper-latin' | 'armenian' | 'georgian' | 'lower-alpha' | 'upper-alpha'
  customBullet?: string;
}

export const List: FC<ListProps> = ({
  items,
  ordered = false,
  nameList,
  disabled = false,
  className = '',
  bullet = 'none',
  customBullet = '',
  unstyled = false,
  ...props
}) => {

  const listStyles = `
    ${styles[!unstyled ? 'list-component' : 'list-unstyled']}
    ${styles[bullet ? bullet : 'none']}
    ${className ? className : ''}
  `

  const Tag: FC<{ children: React.ReactNode }> = ({
    children
  }) => {
    const tagProps = {
      className: listStyles,
      role: "list",
      "aria-label": nameList ? nameList : '',
      ...props
    }

    return ordered
      ? <ol {...tagProps}>{children}</ol>
      : <ul {...tagProps}>{children}</ul>
  }

  return (
    <Tag>
      {items.map((item) => (
        <li
          key={item.id ? item.id : uuid()}
          role="listitem"
          style={customBullet ? { listStyle: customBullet } : {}}>
          {item.text}
        </li>
      ))}
    </Tag>
  )
}

export default List
