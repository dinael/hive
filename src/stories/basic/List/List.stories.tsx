import { List, ListProps } from '../../../components/list/List'

export default {
  title: 'Components/Basic/List',
  component: List,
  parameters: {
    vitest: {
      testFile: "List.test.tsx"
    }
  },
  tags: ['autodocs'],
  argType: {
    nameList: {name: 'List name', control: 'text'},
    items: {name: 'Items', control: 'object' },
    ordered: {name:'Ordered', control: 'boolean'},
    disabled: {name:'Disabled', control: 'boolean'},
    className: { name: 'className', control: 'boolean' },
    bullet: {
      name: 'Bullet',
      control: {
        type: 'select',
        options: [
          'none',
          'auto',
          'disc',
          'circle',
          'square',
          'decimal',
          'decimal-leading-zero',
          'lower-roman',
          'upper-roman',
          'lower-greek',
          'lower-latin',
          'upper-latin',
          'armenian',
          'georgian',
          'lower-alpha',
          'upper-alpha'
        ]
      }
    }
  }
}

const itemsDemo = [
  { id: 'item-1', text: 'Item 1' },
  { id: 'item-2', text: 'Item 2' },
  { id: 'item-3', text: 'Item 3' },
]

// ## Ordered

export const Ordered = (args: ListProps) => (
  [<List {...args} />]
)

Ordered.args = {
  items: itemsDemo,
  nameList: 'Ordered list',
  ordered: true,
  className:'',
  bullet: 'decimal',
}

// ## Unordered
export const Unordered = (args: ListProps) => (
  [<List {...args} />]
)

Unordered.args = {
  ...Ordered.args,
  nameList: 'Unordered list',
  ordered: false,
  bullet: 'circle',
}

// ## None
export const None = (args: ListProps) => (
  [<List {...args} />]
)

None.args = {
  ...Ordered.args,
  nameList: 'None list',
  ordered: false,
  bullet: 'none',
}
