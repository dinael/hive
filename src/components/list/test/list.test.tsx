import { describe, test, expect } from 'vitest'
import { render } from '@testing-library/react'
import List, { ListProps } from '../List'

const listObj = [
  { id: '1', text: 'Item 1' },
  { id: '2', text: 'Item 2' },
  { id: '3', text: 'Item 3' }
]

describe('List', () => {
  const defaultProps: ListProps = {
    bullet: "circle",
    className: "className-test",
    ordered: false,
    items: listObj
  }

  test('renders an unordered list by default', () => {
    const { container } = render(<List {...defaultProps} />)
    const listElement = container.querySelector('ul')
    expect(listElement).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  test('renders an ordered list when `ordered` prop is true', () => {
    const { container } = render(<List {...defaultProps} ordered />)
    const listElement = container.querySelector('ol')
    expect(listElement).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  test('renders the list items', () => {
    const { container } = render(<List {...defaultProps} />)
    const listItems = container.querySelectorAll('li')
    expect(listItems).toHaveLength(defaultProps.items.length)
    expect(listItems[0].textContent).toBe('Item 1')
    expect(listItems[1].textContent).toBe('Item 2')
    expect(listItems[2].textContent).toBe('Item 3')
  })

  test('renders with a custom bullet style', () => {
    const { container } = render(<List {...defaultProps}/>)
    const listElement = container.querySelector('ul')
    expect(listElement?.className).toContain('circle')
  })
  test('renders with a custom class', () => {
    const { container } = render(<List {...defaultProps}/>)
    const listElement = container.querySelector('ul')
    expect(listElement?.className).toContain('className-test')
  })
})
