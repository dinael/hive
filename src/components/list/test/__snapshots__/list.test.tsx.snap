// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List > List renders an ordered list when \`ordered\` prop is true 1`] = `
<div>
  <ol
    aria-label=""
    class="
        list-component
        circle
        className-test"
    role="list"
  >
    <li
      role="listitem"
    >
      Item 1
    </li>
    <li
      role="listitem"
    >
      Item 2
    </li>
    <li
      role="listitem"
    >
      Item 3
    </li>
  </ol>
</div>
`;

exports[`List > List renders an unordered list by default 1`] = `
<div>
  <ul
    aria-label=""
    class="
        list-component
        circle
        className-test"
    role="list"
  >
    <li
      role="listitem"
    >
      Item 1
    </li>
    <li
      role="listitem"
    >
      Item 2
    </li>
    <li
      role="listitem"
    >
      Item 3
    </li>
  </ul>
</div>
`;
